### DISCUSSION TITLE
요청값을 열거체로 파싱하는 경우

### DISCUSSION THREAD

**Discussion by kmw2378** [KakaoFunding/back-end#30] [Ideas]
아래와 같은 JSON을 DTO 클래스로 파싱하려 합니다.
```json
{
    "name" : "lee",
    "gender" : "female"
}
```

`gender`의 경우 female, male 외엔 값 저장이 안되므로 이를 검증하기 위해 별도의 `Gender` 라는 열거체를 사용하기로 했습니다. (스프링 내부적으로 enum converter가 존재하므로)

```java
public enum Gender{
    MALE, FEMALE;
}
```

그러나, 이 `Gender` 라는 열거체는 엔티티 필드에도 사용되어 도메인 패키지에 존재합니다.
```java
@Entity
public class Member {
    ...
    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private Gender gender;
    ...
}
```

따라서 아래와 같이 DTO 필드 타입을 `Gender`로 한다면 엔티티-DTO 간 의존성이 더 강해지는 느낌이 듭니다.
```java
public class ProfileDto {
    private final String name;
    private final Gender gender;
    ...
}
```

추가로 AOP를 통한 예외 핸들링시 DTO 열거체 필드, 나머지 필드로 구분하여 예외 처리를 한 경험이 있어 예외 처리를 통일하기 위해서라도 이 방법을 지양하고 싶었습니다.

그래서 별도의 `@EnumValue` 라는 커스텀 검증 어노테이션을 만들어 기존 Bean Validation 어노테이션 (e.g. `@NotNull`, `@Size`)와 동일하게 예외 핸들링을 하려고 합니다! 

속성으로 파싱할 열거체 클래스, 예외 메시지, 대/소문자 구분 여부를 설정하여 해당 열거체 이름으로 파싱이 가능한지 여부를 검증합니다.

```java
@EnumValue(enumClass = Gender.class, message = "성별이 잘못되었습니다.", ignoreCase = true)
public abstract String getGender();
```

아래는 `@EnumValue`와 Resolver 클래스 내부 코드 입니다.

__EnumValue.java__
```java
@Constraint(validatedBy = EnumValidator.class)
@Target({ElementType.FIELD, ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
public @interface EnumValue {
    Class<? extends Enum<?>> enumClass();
    String message() default "";
    Class<?> [] groups() default {};
    Class<? extends Payload>[] payload() default {};
    boolean ignoreCase() default false;
}
```
__EnumValidator.java__
```java
public class EnumValidator implements ConstraintValidator<EnumValue, String> {
    private EnumValue enumValue;

    @Override
    public void initialize(final EnumValue constraintAnnotation) {
        this.enumValue = enumValue;
    }

    @Override
    public boolean isValid(final String value, final ConstraintValidatorContext context) {
        final Enum<?>[] enumConstants = enumValue.enumClass().getEnumConstants();
        if (enumConstants == null) {
            return false;
        }

        return Arrays.stream(enumConstants)
                .anyMatch(enumConstant -> convertible(value, enumConstant));
    }

    private boolean convertible(final String value, final Enum<?> enumConstant) {
        final String trimValue = value.trim();
        if (enumValue.ignoreCase()) {
            return trimValue.equalsIgnoreCase(enumConstant.name());
        }

        return trimValue.equals(enumConstant.name());
    }
}
```

이 어노테이션을 활용하여 열거체 값을 검증하는건 어떠실까요?? [참고 링크](https://velog.io/@kmw89891/DTO-enum-%ED%95%84%EB%93%9C-%EC%82%AC%EC%9A%A9#string%EC%9C%BC%EB%A1%9C-%EA%B7%B8%EB%83%A5-%EB%91%90%EA%B3%A0-%EA%B2%80%EC%A6%9D-%EC%96%B4%EB%85%B8%ED%85%8C%EC%9D%B4%EC%85%98%EC%9D%84-%EC%82%AC%EC%9A%A9%ED%95%98%EC%9E%90)

**Comment by YeaChan05** [comment]
확실회 엔드포인트단의 열거체 검증 과정이 필요하다는 생각은 있었습니다
제 상황에도 카테고리 기준 조회를 하게될 때 정렬 조건을 열거체를 사용해서 변환하려 했는데, request를 validation하기에 엔드포인트단에 검증 로직이 너무 구구절절 길어질거같아서 일단 문자열로 처리했었는데 저런 방식이라면 마음 놓고 열거체를 사용해도 되겠네요! 저는 너무 좋습니다 ☺️
